src/d3a/models/read_user_profile.py

```diff
@@ -150,8 +146,13 @@ def _calculate_energy_from_power_profile(profile_data_W: Dict[str, float],
     :param slot_length: slot length duration
     :return: a mapping from time to energy values in kWh
     """
-
     input_time_list = list(profile_data_W.keys())
+
+    # add one market slot in order to have another data point for integration
+    additional_time_stamp = input_time_list[-1] + slot_length
+    profile_data_W[additional_time_stamp] = profile_data_W[input_time_list[-1]]
+    input_time_list.append(additional_time_stamp)
```

Comment:
Can you please explain where is this needed ? Should not hurt, just out of curiosity.

Reply:
I was waiting for this question and i though i understood it, but I apparently not. I am currently investigating why i needed this. Will come back to you

Reply:
Ah yes, I was right after all (in my head ;-)) It took me a while now to understand again. This function is really noe easy to understand and i remember long discussions last year about it. 
In short: I is because of the `range()` in line 162.
In length: Due to the nature of `range()`, the last market slot will not be part of the `slot_time_list` (which should be renamed to `slot_time_list_seconds`, if you agree, will do here). Consequently, we need to add another time stamp at the end right from the start. 
If you abstract it into formula (which i did in my head), for calculating the energy you need the power and the time duration. Of course, the time duration should always be the slot length here. But the way the function is implemented, it is calculating the time duration on the fly by difference between the time stamps in order to also take profiles that don't follow the market_slot length. So you need an additional sampling point for that. 
Of course this function could be refactored now, as the `_fill_gaps_in_profile` is called before. But frankly i can not be bothered. Too much time was spent by me in this function already only for this ticket. 
Let me know if my lengthly description doesn't make sense.

Reply:
No that makes sense, thanks for explaining